android/java/org/chromium/chrome/browser/undo_tab_close_snackbar/BraveUndoBarController.java

```diff
@@ -0,0 +1,53 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+package org.chromium.chrome.browser.undo_tab_close_snackbar;
+
+import androidx.annotation.Nullable;
+
+import org.chromium.base.BravePreferenceKeys;
+import org.chromium.chrome.browser.preferences.ChromeSharedPreferences;
+import org.chromium.chrome.browser.tab.Tab;
+import org.chromium.chrome.browser.tabmodel.TabModelSelector;
+import org.chromium.chrome.browser.ui.messages.snackbar.SnackbarManager.SnackbarManageable;
+
+import java.util.List;
+import java.util.function.Supplier;
+
+import javax.naming.Context;
+
+public class BraveUndoBarController extends UndoBarController {
+    public BraveUndoBarController(
+            Context context,
+            TabModelSelector selector,
+            SnackbarManageable snackbarManageable,
+            @Nullable Supplier<Boolean> dialogVisibilitySupplier) {
+        super(context, selector, snackbarManageable, dialogVisibilitySupplier);
+    }
+
+    @Override
+    protected void showUndoBar(List<Tab> closedTabs, boolean isAllTabs) {
+        boolean should_show_undo_bar =
+                ChromeSharedPreferences.getInstance()
+                        .readBoolean(BravePreferenceKeys.SHOW_UNDO_ON_TAB_CLOSED, true);
+        if (!should_show_undo_bar) {
+            if (closedTabs.isEmpty()) {
+                return;
+            }
+            Object actionData = closedTabs.size() == 1 ? closedTabs.get(0).getId() : closedTabs;
+            onDismissNoAction(actionData);
+            return;
+        }
+
+        BraveReflectionUtil.InvokeMethod(
+                UndoBarController.class,
+                this,
+                "showUndoBar",
+                List.class,
+                closedTabs,
+                boolean.class,
+                isAllTabs);
```

Comment:
I've actually tried that, but encountered the following error during build time, which is why I thought of switching to reflection.

```
../../brave/android/java/org/chromium/chrome/browser/undo_tab_close_snackbar/BraveUndoBarController.java:45: 
error: showUndoBar(List<Tab>,boolean) has private access in UndoBarController
        super.showUndoBar(closedTabs, isAllTabs);
```

Btw, `BraveClassAdapter` will only get executed after compilation right(i.e after javac) ?

@SergeyZhukovsky

Reply:
yep it generates everything after.
Actually you are right you need to call reflection there, please check how we do it with that method https://github.com/brave/brave-core/blob/master/build/android/bytecode/java/org/brave/bytecode/BraveTabbedActivityClassAdapter.java#L54. You have to redirect an ownership.

Reply:
I've tried that as well and encountered the same error.

I have a doubt - if the Adapter processes happen only after compilation, how would they help resolve the error we're encountering during compilation, is there any catch there?
@SergeyZhukovsky

Reply:
One more note. In order reflection calls to work in release builds we need to add proguard rule to avoid this method obfuscation. Otherwise it won't be able to find this method. Here is how ti's done for the example above https://github.com/brave/brave-core/blob/master/android/java/proguard.flags#L68

Reply:
Thanks @samartnik! I'll  check this!